### PR TITLE
feat(farm): extending from config testnet

### PR DESCRIPTION
**Pull request by NikolaMilosa**
Support regular `Subnet` and `Node` structures along with the following:
```json
{
  "subnets": [
    {
      "subnet_type": "application",
      "num_nodes": 2
    },
    {
      "subnet_type": "system",
      "num_nodes": 3
    }
  ],
  "num_unassigned_nodes": 0,
  "initial_version": "a3831c87440df4821b435050c8a8fcb3745d86f6"
}

```

### CODE REVIEW

File: rs/tests/driver/Cargo.toml [hunk 1]
```diff
@@ -130,7 +131,7 @@ serde_bytes = { workspace = true }
 serde_cbor = { workspace = true }
 serde_json = { workspace = true }
 serde_yaml = { workspace = true }
-sha2 = { workspace = true }
+sha2 = "0.10"
```
Comment (basvandijk) [inline rs/tests/driver/Cargo.toml:134]: Can't we keep this at the workspace version?

**Review by basvandijk — COMMENTED** [review]
Great stuff!

But are you sure this works correctly? I would have expected that you also need to overwrite the [ic_os_img_sha256 and ic_os_img_url](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/driver/src/driver/resource.rs?L158) with the correct URL and hash of the disk image (i.e. not the update image).

File: rs/tests/driver/Cargo.toml (hunk 1 above)
Comment (NikolaMilosa) [reply rs/tests/driver/Cargo.toml]: Fixed, we should use the same one, my bad

**Comment by NikolaMilosa** [comment]
From my testing i've seen that If i include the url and hash in [there](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/driver/src/driver/resource.rs?L158) Farm panics saying that it cannot find the image in farm artifacts. With the code I did we get a testnet with `0000...` version and then it progresses to specified version in the config.

**Comment by basvandijk** [comment]
> With the code I did we get a testnet with 0000... version and then it progresses to specified version in the config.

Yes, it will perform an upgrade from 0000... to the version specified. But that should be unnecessarry if you already start from the right version.

**Comment by basvandijk** [comment]
So you should consider overriding `ENV_DEPS__DEV_DISK_IMG_TAR_ZST_CAS_URL` and `ENV_DEPS__DEV_DISK_IMG_TAR_ZST_SHA256` the same way you do for the update image.

File: rs/tests/testing_verification/testnets/from_config.rs [hunk 2]
```diff
@@ -52,34 +107,111 @@ pub fn setup(env: TestEnv) {
...
+            &url, TAR_EXTENSION
+        )
+    })
+    .split_whitespace()
+    .into_iter()
+    .next()
+    .expect("The format of hash should contain whitespace")
+    .to_string()
+}
+
```
Comment (NikolaMilosa) [inline rs/tests/testing_verification/testnets/from_config.rs:178]: Managed to do it how you suggested. The good thing is that fetching `sha` will fail if the version is not present so we cover that path as well 💯

File: rs/tests/driver/Cargo.toml [hunk 3]
```diff
@@ -27,6 +27,7 @@ ed25519-dalek = { workspace = true }
 flate2 = { workspace = true }
 futures = { workspace = true }
 humantime = "2.0"
+humantime-serde = "1.1.1"
```
Comment (basvandijk) [inline rs/tests/driver/Cargo.toml:30]: Could you add this to the workspace? We recently added a [Rust Dependency Policy](https://github.com/dfinity/ic/blob/master/README.adoc#rust-dependency-policy) that requires this.

File: rs/tests/testing_verification/testnets/from_config.rs [hunk 4]
```diff
@@ -52,34 +107,111 @@ pub fn setup(env: TestEnv) {
...
     }
 
     env.sync_with_prometheus();
 }
 
+fn write_file_and_update_env_variable(env: &TestEnv, pairs: Vec<(&str, String, &str)>) {
+    for (file_name, value_in_file, env_variable) in pairs {
+        let path = env.get_dependency_path(file_name);
+        std::fs::write(&path, value_in_file)
+            .unwrap_or_else(|_| panic!("Failed to write to path: {}", path.display()));
```
Comment (basvandijk) [inline rs/tests/testing_verification/testnets/from_config.rs:131]: What's the purpose of writing these files?

Comment (NikolaMilosa) [reply rs/tests/testing_verification/testnets/from_config.rs:131]: environment values only give the path to files which contain information. That is how `HasDependencies`  trait is implemented

**Review by sasa-tomic — APPROVED** [review]
I'm not an expert in the code, but from a non-expert look - LGTM!

**Pull request merged**